spot-contracts/contracts/PerpetualTranche.sol

```diff
@@ -887,17 +897,43 @@ contract PerpetualTranche is
     }
 
     /// @dev Enforces the total supply and per tranche mint cap. To be invoked AFTER the mint operation.
-    function _enforceMintCaps(ITranche trancheIn) private view {
-        // checks if supply minted using the given tranche is within the cap
-        if (mintedSupplyPerTranche[trancheIn] > maxMintAmtPerTranche) {
-            revert ExceededMaxMintPerTranche();
-        }
-
-        // checks if new total supply is within the max supply cap
+    function _enforceMintCaps(ITranche depositTranche) private view {
+        // Checks if new total supply is within the max supply cap
         uint256 newSupply = totalSupply();
         if (newSupply > maxSupply) {
             revert ExceededMaxSupply();
         }
+
+        // Checks if the value of deposit tranche relative to the other tranches in the reserve
+        // is no higher than the defined limit.
+        //
+        // NOTE: We consider the tranches which are up for rollover,
+        // to be part of the deposit tranche, as given enough time
+        // they will be eventually rolled over into the deposit tranche.
+        // If the reserve has any mature underlying tokens,
+        // we just exempt it from our calculation.
+        IERC20Upgradeable underlying_ = _reserveAt(0);
+        uint256 totalTrancheVal = 0;
+        uint256 depositTrancheValue = 0;
+        uint8 reserveCount = uint8(_reserves.length());
+        for (uint8 i = 1; i < reserveCount; ++i) {
+            ITranche tranche = ITranche(address(_reserveAt(i)));
+            uint256 trancheValue = _computeReserveTrancheValue(
```

Comment:
I was trying to think whether we should be using tranche values or tranche balances...

I think I agree with using values--it's the more conservative one, though, since it would cut off minting earlier. It would naturally get relaxed as later tranches get closer to being overwater again.